### PR TITLE
Network initial work

### PR DESCRIPTION
**Pull request by osmirnov**
What is done:
- Introduced support multiple network transports
- Changed configuration to support initial peers list
- Moved TCP specific stuff out from Server into TcpPeer (planning WsPeer, not sure about HttpPeer)
- Introduced some Task-like workflows

### CODE REVIEW

File: src/NeoSharp.Core/Network/EndPoint.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿namespace NeoSharp.Core.Network
+{
+    public class EndPoint
+    {
+        public Protocol Protocol { get; set; }
+
+        public string Host { get; set; }
```
Comment (shargon) [inline src/NeoSharp.Core/Network/EndPoint.cs]: i prefer IpAddress or IpEndPoint here

File: src/NeoSharp.Core/Network/Server.cs [hunk 2]
```diff
@@ -10,60 +13,65 @@ public class Server : IServer
...
-        
+        private readonly IPeerFactory _peerFactory;
+
         // ReSharper disable once FieldCanBeMadeReadOnly.Local
         // ReSharper disable once CollectionNeverUpdated.Local
-        private IList<IPeer> _connectedPeers;                // if we successfully connect with a peer it is inserted into this list
+        private ConcurrentBag<IPeer> _connectedPeers;                // if we successfully connect with a peer it is inserted into this list
         // ReSharper disable once NotAccessedField.Local
         private IList<IPEndPoint> _failedPeers;             // if we can't connect to a peer it is inserted into this list
         private uint _nonce;                                // uniquely identifies this server so we can filter out our own messages sent back to us by other nodes
```
Comment (shargon) [inline src/NeoSharp.Core/Network/Server.cs]: I prefer Guid for next version of protocol, is more unique, maybe i should add this change to https://github.com/neo-project/proposals/pull/31

File: src/NeoSharp.Core/Network/NetworkConfig.cs [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
-        public string ServerIp { get; set; }
+        private static readonly Regex _peerEndPointPattern = new Regex(@"^(?<proto>\w+)://(?<host>[^/:]+):?(?<port>\d+)?/?$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
 
-        public int ServerStartPort { get; set; }
+        public EndPoint[] PeerEndPoints { get; private set; }
 
         public NetworkConfig(IConfiguration configuration = null)
         {
-            configuration?.GetSection(nameof(NetworkConfig))?.Bind(this);
+            var peerEndPoints = configuration?.GetSection(nameof(NetworkConfig))?.GetSection("PeerEndPoints").Get<string[]>();
```
Comment (shargon) [inline src/NeoSharp.Core/Network/NetworkConfig.cs]: Add null check at the end if is necessary please
ˋ?.Get<string[]>();ˋ

File: src/NeoSharp.Core/Network/EndPoint.cs (hunk 1 above)
Comment (osmirnov) [reply src/NeoSharp.Core/Network/EndPoint.cs]: Host property can hold name or IP address. I've checked the previous implementation and found WebSocket does not use either IPEndPoint or IPAddress (https://github.com/neo-project/neo/blob/11d8db11568d9eadeeb86c5b8c21a1d3937e0912/neo/Network/LocalNode.cs#L632). I will reconsider this choice when I implement TCP communication and DNS IP resolution. Thanks for the suggestion.

File: src/NeoSharp.Core/Network/Server.cs (hunk 2 above)
Comment (osmirnov) [reply src/NeoSharp.Core/Network/Server.cs]: I am not familiar with the protocol yet. Will it make this incompatible with old one?

File: src/NeoSharp.Core/Network/NetworkConfig.cs (hunk 3 above)
Comment (osmirnov) [reply src/NeoSharp.Core/Network/NetworkConfig.cs]: Good point. All this ? in one line smell a bit.

**Comment by gubanotorious** [comment]
We can merge for now and continue to refine as we move further down the path of implementation

**Pull request merged**